### PR TITLE
Added custom authentication support to Garnet.

### PR DESCRIPTION
**Pull request by Xizt**
Changes to support custom authentication. Extended AuthenticationSettings by adding a custom mode which is expected to be passed during server start if custom auth mode is chosen.

### CODE REVIEW

**Comment by Xizt** [comment]
@microsoft-github-policy-service agree [company="microsoft"]

File: libs/host/Configuration/Options.cs [hunk 1]
```diff
@@ -493,7 +493,7 @@ public bool IsValid(out List<string> invalidOptions, ILogger logger)
             return isValid;
         }
 
-        public GarnetServerOptions GetServerOptions(ILogger logger = null)
+        public GarnetServerOptions GetServerOptions(IAuthenticationSettings authenticationSettings = null, ILogger logger = null)
```
Comment (vazois) [inline libs/host/Configuration/Options.cs:496]: I am not sure I understand why we need this. Instead of passing authenticationSettings as a parameter one could do the following if they need custom authenticationSettings

this.opts = serverSettings.GetServerOptions(this.loggerFactory.CreateLogger("Options"));
this.opts.AuthSettings = new CustomAuthenticationSettings();

File: libs/server/Auth/Settings/AuthenticationSettings.cs [hunk 2]
```diff
@@ -34,7 +34,12 @@ public enum GarnetAuthenticationMode
         /// <summary>
         /// ACL mode using Aad token instead of password. Here username is expected to be ObjectId or a valid Group's Object Id and token will be validated for claims.
         /// </summary>
-        AclWithAad
+        AclWithAad,
+
+        /// <summary>
+        /// Custom authentication mode. Requires a custom authenticator to be implemented and passed during Garnet server instantiation.
+        /// </summary>
+        Custom
```
Comment (vazois) [inline libs/server/Auth/Settings/AuthenticationSettings.cs:42]: I don't think we should expose the custom option, because the user cannot provide an IAuthenticationSettings command line parameter. It will be confusing and may create bugs.

File: libs/host/Configuration/Options.cs [hunk 3]
```diff
@@ -663,6 +663,14 @@ private IAuthenticationSettings GetAuthenticationSettings(ILogger logger = null)
                 case GarnetAuthenticationMode.AclWithAad:
                     var aadAuthSettings = new AadAuthenticationSettings(AuthorizedAadApplicationIds?.Split(','), AadAudiences?.Split(','), AadIssuers?.Split(','), IssuerSigningTokenProvider.Create(AadAuthority, logger), AadValidateUsername.GetValueOrDefault());
                     return new AclAuthenticationAadSettings(AclFile, Password, aadAuthSettings);
+                case GarnetAuthenticationMode.Custom:
```
Comment (vazois) [inline libs/host/Configuration/Options.cs:666]: maybe we could default this to NoAuth and print a warning message. The caller is responsible for providing authentication settings. It might be better than passing it as a parameter.

**Review by vazois — CHANGES_REQUESTED** [review]
Need some changes before merging

File: libs/host/Configuration/Options.cs (hunk 1 above)
Comment (badrishc) [reply libs/host/Configuration/Options.cs:496]: +1

File: libs/host/GarnetServer.cs [hunk 4]
```diff
@@ -76,7 +77,7 @@ public class GarnetServer : IDisposable
         /// </summary>
         /// <param name="commandLineArgs">Command line arguments</param>
         /// <param name="loggerFactory">Logger factory</param>
-        public GarnetServer(string[] commandLineArgs, ILoggerFactory loggerFactory = null, bool cleanupDir = false)
+        public GarnetServer(string[] commandLineArgs, IAuthenticationSettings authenticationSettings = null, ILoggerFactory loggerFactory = null, bool cleanupDir = false)
```
Comment (badrishc) [inline libs/host/GarnetServer.cs:80]: Rename to authenticationSettingsOverride and use that if not null, instead of creating a new enum type.

File: libs/host/Configuration/Options.cs (hunk 1 above)
Comment (Xizt) [reply libs/host/Configuration/Options.cs]: @vazois , Updated.

File: libs/host/Configuration/Options.cs (hunk 3 above)
Comment (Xizt) [reply libs/host/Configuration/Options.cs]: Updated

File: libs/host/Configuration/Options.cs [hunk 5]
```diff
@@ -664,8 +664,8 @@ private IAuthenticationSettings GetAuthenticationSettings(ILogger logger = null)
                     var aadAuthSettings = new AadAuthenticationSettings(AuthorizedAadApplicationIds?.Split(','), AadAudiences?.Split(','), AadIssuers?.Split(','), IssuerSigningTokenProvider.Create(AadAuthority, logger), AadValidateUsername.GetValueOrDefault());
                     return new AclAuthenticationAadSettings(AclFile, Password, aadAuthSettings);
                 default:
-                    logger?.LogError("Unsupported authentication mode: {mode}", AuthenticationMode);
-                    throw new Exception($"Authentication mode {AuthenticationMode} is not supported.");
+                    logger?.LogWarning("Defaulting to NoAuth if authenticationSettingsOverride is not provided.");
```
Comment (badrishc) [inline libs/host/Configuration/Options.cs:667]: why is this needed? you can send NoAuth and provide override.

Comment (Xizt) [reply libs/host/Configuration/Options.cs]: Resolved, Thanks

File: libs/host/GarnetServer.cs [hunk 6]
```diff
@@ -76,7 +77,7 @@ public class GarnetServer : IDisposable
         /// </summary>
         /// <param name="commandLineArgs">Command line arguments</param>
         /// <param name="loggerFactory">Logger factory</param>
-        public GarnetServer(string[] commandLineArgs, ILoggerFactory loggerFactory = null, bool cleanupDir = false)
+        public GarnetServer(string[] commandLineArgs, IAuthenticationSettings authenticationSettingsOverride = null, ILoggerFactory loggerFactory = null, bool cleanupDir = false)
```
Comment (msft-paddy14) [inline libs/host/GarnetServer.cs:80]: should we not create a new constructor for this or add it at the end atleast?
This will break for anyone already using existing optional parameters ( ex: an ILoggerFactory instance as 2nd parameter)

Comment (Xizt) [reply libs/host/GarnetServer.cs]: Thanks. Need to keep in mind for compatibility.

**Review by badrishc — APPROVED** [review]

**Review by vazois — APPROVED** [review]
